<a href="https://colab.research.google.com/github/mario-rot/Introduction-to-Human-Language-Technology/blob/main/Session6_LaurenTucker_MarioRosas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab session 6 (Word Sense Disambiguation) - ILTH

**Students:** Lauren Tucker & Mario Rosas

## Paraphrases Template

In [1]:
%%shell
git clone https://github.com/mario-rot/Introduction-to-Human-Language-Technology.git
cd 'Introduction-to-Human-Language-Technology'
mv 'Complementary Material' /content/

Cloning into 'Introduction-to-Human-Language-Technology'...
remote: Enumerating objects: 849, done.
remote: Counting objects: 100% (849/849), done.
remote: Compressing objects: 100% (777/777), done.
remote: Total 849 (delta 107), reused 788 (delta 67), pack-reused 0
Receiving objects: 100% (849/849), 2.25 MiB | 17.73 MiB/s, done.
Resolving deltas: 100% (107/107), done.


In [1]:
import pandas as pd

In [2]:
dt = pd.read_csv('Complementary Material/test-gold/STS.input.SMTeuroparl.txt',sep='\t',header=None)

In [3]:
dt.head()

0  \
0  The leaders have now been given a new chance a...   
1  Amendment No 7 proposes certain changes in the...   
2  Let me remind you that our allies include ferv...   
3        The vote will take place today at 5.30 p.m.   
4  The fishermen are inactive, tired and disappoi...   

                                                   1  
0  The leaders benefit aujourd' hui of a new luck...  
1  Amendment No 7 is proposing certain changes in...  
2  I would like to remind you that among our alli...  
3                 The vote will take place at 5.30pm  
4  The fishermen are inactive, tired and disappoi...

In [4]:
dt['gs'] = pd.read_csv('Complementary Material/test-gold/STS.gs.SMTeuroparl.txt',sep='\t',header=None)

In [5]:
dt.shape

(459, 3)

In [6]:
dt.head()

0  \
0  The leaders have now been given a new chance a...   
1  Amendment No 7 proposes certain changes in the...   
2  Let me remind you that our allies include ferv...   
3        The vote will take place today at 5.30 p.m.   
4  The fishermen are inactive, tired and disappoi...   

                                                   1    gs  
0  The leaders benefit aujourd' hui of a new luck...  4.50  
1  Amendment No 7 is proposing certain changes in...  5.00  
2  I would like to remind you that among our alli...  4.25  
3                 The vote will take place at 5.30pm  4.50  
4  The fishermen are inactive, tired and disappoi...  5.00

# Excercise

In [7]:
from nltk.metrics import jaccard_distance
from nltk.stem import WordNetLemmatizer
from nltk import pos_tag
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
nltk.download('omw-1.4')
from scipy.stats import pearsonr
import numpy as np

import string
import collections
import nltk
from nltk.corpus.reader.wordnet import VERB, NOUN, ADJ, ADV
from nltk.corpus import gutenberg
from nltk.text import Text
nltk.download('gutenberg')
nltk.download('stopwords')

def remove_signs(wrd):
  wrd = list(wrd)
  wrd = [word for word in wrd if not any(caracter in signs for caracter in word)]
  wrd = ''.join(wrd)
  return wrd

def clean(corpus, stopwords, minwords_len, signs):
   corpus = corpus.split(' ')
   corpus = [word.lower() for word in corpus]
   corpus = [word if not any(caracter in signs for caracter in word) else remove_signs(word) for word in corpus]
   corpus = [word for word in corpus if word not in stopwords and word.isalpha()]
   corpus = [word for word in corpus if len(word) > minwords_len]
   return corpus

from nltk.corpus.reader.wordnet import VERB, NOUN, ADJ, ADV
pos_map = {'N': NOUN,
           'V':VERB,
           'J':ADJ,
           'R':ADV}

# corpus = list(gutenberg.words('blake-poems.txt'))
stopwords=set(nltk.corpus.stopwords.words('english'))
signs = string.punctuation
minwords_len = 2

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\mario\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\mario\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\mario\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\mario\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package gutenberg to
[nltk_data]     C:\Users\mario\AppData\Roaming\nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\mario\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [8]:
cleaned = clean(dt.loc[0][0], stopwords, minwords_len, signs)
cleaned

['leaders', 'given', 'new', 'chance', 'let', 'hope', 'seize']

In [9]:
accepted_pos = ['JJ', 'JJR', 'JJS', 'NN', 'NNS', 'NNP', 'NNPS', 'RB', 'RBR', 'RBS', 'VB', 'VBD', 'VBN', 'VBP', 'VBZ']

pos_map = {'n': ['NN', 'NNS', 'NNP', 'NNPS'],
           'v': ['VB', 'VBD', 'VBN', 'VBP', 'VBZ'],
           'j': ['JJ', 'JJR', 'JJS'],
           'r': ['RB', 'RBR', 'RBS']}
correcting = {'n':'n', 'v':'v', 'j':'a', 'r':'r'}
pos_map2 = {'N': NOUN,
           'V':VERB,
           'J':ADJ,
           'R':ADV}


def filter_pos(pair):
    if pair[1][0].lower() in list(pos_map.keys()):
        return pair[0], correcting[pair[1][0].lower()]
    return None, None

# def filter_pos(pair):
#     if pair[1] in accepted_pos:
#         return pair
#     return None, None

In [80]:
def applying_lesk(sentence):
    cleaned = clean(sentence, stopwords, minwords_len, signs)
    pairs = nltk.pos_tag(cleaned)
    synsets = []
    for pair in pairs:
      context = cleaned
      word, pos = filter_pos(pair)
      if pos:
        synset = nltk.wsd.lesk(context, word, pos)
      else:
          synset = False
      if synset:
        synsets.append(synset)
    return synsets


In [69]:
wnl = nltk.stem.WordNetLemmatizer()


def lemmatize(p):
    if p[1][0] in {'N', 'V', 'J', 'R'}:
        return wnl.lemmatize(p[0].lower(), pos=pos_map2[p[1][0]])
    return p[0]


def lemmatize_sentence(sentence):
    cleaned = clean(sentence, stopwords, minwords_len, signs)
    tagged = nltk.pos_tag(cleaned)
    return [[lemmatize(pair),pair[1]] for pair in tagged]


def applying_lesk_lemmas(sentence):
    context = [i[0] for i in sentence]
    synsets = []
    for pair in sentence:
        context = cleaned
        word, pos = filter_pos(pair)
        if pos:
            synset = nltk.wsd.lesk(context, word, pos)
        else:
            synset = False
        if synset:
            synsets.append(synset)
    return synsets

In [12]:
def compute_metric(metric, elements): # Fenction to calculate Jaccard Distance
    if metric == 'jaccard':
        res = jaccard_distance(set(applying_lesk(elements[0])),
                               set(applying_lesk(elements[1])))
    if metric == 'jaccard_lemmas':
      res = jaccard_distance(set(applying_lesk_lemmas(lemmatize_sentence(elements[0]))),
                             set(applying_lesk_lemmas(lemmatize_sentence(elements[1]))))
    return res

In [13]:
dt['jaccard'],dt['jaccard_lemmas'] = "",""

In [14]:
for i in range(len(dt)): # Iterating to get Jaccard distances over each column of the dataframe
  dt['jaccard'][i] = np.float64(compute_metric('jaccard',[dt.iloc[i][0],dt.iloc[i][1]]))*10
  dt['jaccard_lemmas'][i] = np.float64(compute_metric('jaccard_lemmas',[dt.iloc[i][0],dt.iloc[i][1]]))*10

C:\Users\mario\AppData\Local\Temp\ipykernel_12320\3921866450.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt['jaccard'][i] = np.float64(compute_metric('jaccard',[dt.iloc[i][0],dt.iloc[i][1]]))*10
C:\Users\mario\AppData\Local\Temp\ipykernel_12320\3921866450.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt['jaccard_lemmas'][i] = np.float64(compute_metric('jaccard_lemmas',[dt.iloc[i][0],dt.iloc[i][1]]))*10


In [15]:
dt

0  \
0    The leaders have now been given a new chance a...   
1    Amendment No 7 proposes certain changes in the...   
2    Let me remind you that our allies include ferv...   
3          The vote will take place today at 5.30 p.m.   
4    The fishermen are inactive, tired and disappoi...   
..                                                 ...   
454  It is our job to continue to support Latvia wi...   
455        The vote will take place today at 5.30 p.m.   
456  Neither was there a qualified majority within ...   
457  Let me remind you that our allies include ferv...   
458  We often pontificate here about being the repr...   

                                                     1     gs   jaccard  \
0    The leaders benefit aujourd' hui of a new luck...  4.500  5.555556   
1    Amendment No 7 is proposing certain changes in...  5.000       0.0   
2    I would like to remind you that among our alli...  4.250       7.5   
3                   The vote will take place at 5.30pm  4.500       2.5   
4    The fishermen are inactive, tired and disappoi...  5.000       0.0   
..                                                 ...    ...       ...   
454  It is of our duty of continue to support the c...  5.000  6.363636   
455                   Vote will take place at 17 h 30.  4.750       2.5   
456  There was no qualified majority in this Parlia...  5.000       5.0   
457  I hold you recall that our allies, there are e...  4.000       7.5   
458  We often take pride here to represent the citi...  3.833       5.0   

    jaccard_lemmas  
0         5.555556  
1              0.0  
2              7.5  
3              2.5  
4              0.0  
..             ...  
454       6.363636  
455            2.5  
456            5.0  
457            7.5  
458            5.0  

[459 rows x 5 columns]

In [16]:
pearsonr(dt['gs'], 1-dt['jaccard']) # Calculating the pearson correlation between GS results and 1-Jaccard calculated data with lesk algorithm applied

PearsonRResult(statistic=0.454977041340358, pvalue=7.827296967470474e-25)

In [17]:
pearsonr(dt['gs'], 1-dt['jaccard_lemmas']) # Calculating the pearson correlation between GS results and 1-Jaccard calculated data with lesk algorithm + lemmatizer applied

PearsonRResult(statistic=0.4565851262924598, pvalue=5.110934549149753e-25)

In [18]:
count = 0
for index, (j,jl) in enumerate(zip(dt['jaccard'],dt['jaccard_lemmas'])):
    if j != jl:
        count += 1
        print(index,'-->', j,'--',jl)

8 --> 7.777777777777778 -- 6.25
14 --> 10.0 -- 8.75
62 --> 10.0 -- 8.75
80 --> 10.0 -- 8.75
100 --> 7.0 -- 5.555555555555555
118 --> 8.0 -- 6.666666666666666
175 --> 10.0 -- 5.0
184 --> 5.555555555555555 -- 3.75
187 --> 5.7142857142857135 -- 3.333333333333333
188 --> 6.25 -- 4.285714285714286
209 --> 6.666666666666666 -- 0.0
235 --> 7.777777777777778 -- 6.25
236 --> 10.0 -- 8.88888888888889
242 --> 7.0 -- 5.555555555555555
255 --> 8.0 -- 6.666666666666666
256 --> 7.0 -- 5.555555555555555
265 --> 6.666666666666666 -- 5.0
285 --> 2.8571428571428568 -- 0.0
287 --> 10.0 -- 5.0
323 --> 7.777777777777778 -- 6.25
343 --> 5.0 -- 2.8571428571428568
347 --> 6.25 -- 4.285714285714286
404 --> 10.0 -- 8.88888888888889
410 --> 4.285714285714286 -- 1.6666666666666665


# Conclusion
For reference, the Pearson correlation for Lab 2 was:

*PearsonRResult(statistic=0.45049771693186835, pvalue=2.5356459143049236e-24)*

and the Pearson correlation for Lab 3 was:

*PearsonRResult(statistic=0.48102317341708245, pvalue=5.904510415498371e-28)*


In this lab, we first removed stopwords and punctuation from our original sentences and filtered with pos_tag to receive pairs of significant words and their corresponding parts of speech for all nouns, adjectives, adverbs, and verbs. We then applied the lesk algorithm to this cleaned set of words and received a Pearson correlation value of:

*PearsonRResult(statistic=0.454977041340358, pvalue=7.827296967470474e-25)*

We then took an additional step and lemmatized the cleaned set of words and performed lesk on this lemmatized set, and we received a Pearson correlation value of:

*PearsonRResult(statistic=0.4565851262924598, pvalue=5.110934549149753e-25)*


Lesk, both when performed on the lemmatized set and the unlemmatized set, performed better than the document analysis from Lab 2. In both cases, this is likely because the context that the Lesk algorithm uses is supporting the evaluation of word similarities and creates a more accurate picture of the meaning of each word. Lesk occasionally performed better on the lemmatized set than on the unlemmatized set, with 24 of the instances having lower Jaccard coefficients for the lemmatized sets. Contrary to our expectations, the lemmatization does not seem to have a negative impact on the context. In other words, lemmatizing the sentence does not strip any meaning from the context. In any case, however, the improvement in performance when using Lesk is very small, despite our predictions that the improvement would be much greater, since we hypothesized that having context would improve performance.


Lesk, both when performed on the lemmatized set and the unlemmatized set, performed worse than just lemmatizing the words as done in Lab 3, as shown above. One possible reason for this may be that, when taking more words into account when determining similarity (i.e. by adding the context), there is more room for error. If the wrong synset is determined for some word in the context, then it will impact the interpreted meaning of the whole sentence. With just lemmatizing, on the other hand, as done in Lab 3, there is less chance of error, since only the structure of each word is being changed and examined, without the subjective interpretations of meaning used with context analysis.









In [101]:
def applying_lesk(sentence, all = False):
    context = sentence.split(' ')
    cleaned = clean(sentence, stopwords, minwords_len, signs)
    pairs = nltk.pos_tag(cleaned)
    synsets = []
    for pair in pairs:
    #   context = cleaned
        word, pos = filter_pos(pair)
        if pos:
            synset = nltk.wsd.lesk(context, word, pos)
        else:
            synset = False
        if synset:
            synsets.append(synset.lemmas()[0].name())
        elif not synset and all:
            synsets.append(pair[0])
    return synsets

wnl = nltk.stem.WordNetLemmatizer()


def lemmatize(p):
    if p[1][0] in {'N', 'V', 'J', 'R'}:
        return wnl.lemmatize(p[0].lower(), pos=pos_map2[p[1][0]])
    return p[0]


def lemmatize_sentence(sentence):
    cleaned = clean(sentence, stopwords, minwords_len, signs)
    tagged = nltk.pos_tag(cleaned)
    return [[lemmatize(pair),pair[1]] for pair in tagged]


def applying_lesk_lemmas(sentence, all = False):
    context = sentence.split(' ')
    lemmatized_sentence = lemmatize_sentence(sentence)
    synsets = []
    for pair in lemmatized_sentence:
        # context = cleaned
        word, pos = filter_pos(pair)
        if pos:
            synset = nltk.wsd.lesk(context, word, pos)
        else:
            synset = False
        if synset:
            synsets.append(synset.lemmas()[0].name())
        elif not synset and all:
            synsets.append(pair[0])
    return synsets

In [20]:
nltk.word_tokenize(dt[0][0])

['The',
 'leaders',
 'have',
 'now',
 'been',
 'given',
 'a',
 'new',
 'chance',
 'and',
 'let',
 'us',
 'hope',
 'they',
 'seize',
 'it',
 '.']

In [21]:
lemmatize_sentence(dt[0][0])

[['leader', 'NNS'],
 ['give', 'VBN'],
 ['new', 'JJ'],
 ['chance', 'NN'],
 ['let', 'VBD'],
 ['hope', 'PRP'],
 ['seize', 'VB']]

In [76]:
applying_lesk_lemmas(lemmatize_sentence(dt[0][0]))

AttributeError: 'list' object has no attribute 'split'

In [23]:
filter_pos(nltk.pos_tag(clean('neither', stopwords, minwords_len, signs))[0])

(None, None)

In [24]:
from collections import Counter

In [25]:
context = ['I', 'went', 'to', 'the', 'bank', 'to', 'deposit', 'money', '.']

synset = nltk.wsd.lesk(context, 'bank', 'n')

synset.lemmas()[0].name()

'savings_bank'

In [26]:
sen = ' '.join(['I', 'went', 'to', 'the', 'bank', 'to', 'deposit', 'money', '.'])
sen

'I went to the bank to deposit money .'

In [27]:
applying_lesk_lemmas(sen, True)

['survive', 'savings_bank', 'down_payment', 'money']

In [28]:
applying_lesk(sen, True)

['survive', 'savings_bank', 'down_payment', 'money']

In [29]:
sen

'I went to the bank to deposit money .'

In [30]:
nltk.word_tokenize(sen)

['I', 'went', 'to', 'the', 'bank', 'to', 'deposit', 'money', '.']

In [33]:
nltk.pos_tag(nltk.word_tokenize(sen))

[('I', 'PRP'),
 ('went', 'VBD'),
 ('to', 'TO'),
 ('the', 'DT'),
 ('bank', 'NN'),
 ('to', 'TO'),
 ('deposit', 'VB'),
 ('money', 'NN'),
 ('.', '.')]

In [37]:
def applying_lesk(sentence, all = False):
    # context = sentence.split(' ')
    cleaned = nltk.word_tokenize(sentence)
    pairs = nltk.pos_tag(cleaned)
    synsets = []
    for pair in pairs:
    #   context = cleaned
        word, pos = filter_pos(pair)
        if pos:
            synset = nltk.wsd.lesk(cleaned, word, pos).lemmas()[0].name()
        else:
            synset = False
        if synset:
            synsets.append(synset)
        elif not synset and all:
            synsets.append(pair[0])
    return synsets

In [40]:
applying_lesk(sen, True)

['I', 'survive', 'to', 'the', 'savings_bank', 'to', 'deposit', 'money', '.']

In [55]:
print(dt[0][0])
print(dt[1][0])

The leaders have now been given a new chance and let us hope they seize it.
The leaders benefit aujourd' hui of a new luck and let's let them therefore seize it.


In [102]:
applying_lesk(dt[0][0])

['leadership', 'give', 'new', 'probability', 'let', 'assume']

In [103]:
applying_lesk(dt[1][0])

['leadership',
 'profit',
 'new',
 'Lashkar-e-Taiba',
 'let',
 'therefore',
 'impound']

In [83]:
wnl = nltk.stem.WordNetLemmatizer()

In [86]:
from nltk.corpus import wordnet as wn

In [94]:
nltk.pos_tag(['aujourd'])

[('aujourd', 'NN')]

In [95]:
wn.synsets('aujourd', 'n')

[]